gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,272 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *******************************************************************/
+#include <config.h>
+
+#include <gtk/gtk.h>
+#include <glib/gi18n.h>
+#include <vector>
+#include <unordered_map>
+
+#include "inttypes.h"
+#include "Account.h"
+#include "Split.h"
+#include "Transaction.h"
+#include "gncOwner.h"
+#include "qof.h"
+#include "gnc-autoclear.h"
+
+static QofLogModule log_module = GNC_MOD_GUI;
+
+/* the following functions are used in window-autoclear: */
+
+typedef enum
+{
+    AUTOCLEAR_OVERLOAD = 1,
+    AUTOCLEAR_UNABLE,
+    AUTOCLEAR_MULTIPLE,
+    AUTOCLEAR_NOP,
+} autoclear_error_type;
+
+#define UPDATES_PER_SECOND 5
+#define MAX_AUTOCLEAR_SECONDS 10
+
+#define MAX_UPDATE_TICKS CLOCKS_PER_SEC / UPDATES_PER_SECOND
+#define MAX_AUTOCLEAR_TICKS CLOCKS_PER_SEC * MAX_AUTOCLEAR_SECONDS
+
+struct WorkItem
+{
+    gint64 reachable_amount;
+    std::vector<Split*> splits_vector;
+    WorkItem (const gint64 amount, std::vector<Split*> splits)
+        : reachable_amount (amount), splits_vector(splits){}
+};
+
+static void status_update (GtkLabel *label, gchar *status)
+{
+    if (!label) return;
+    gtk_label_set_text (label, status);
+    while (gtk_events_pending ())
+        g_main_context_iteration (nullptr, FALSE);
+}
+
+static void looping_update_status (GtkLabel *label, guint nc_progress,
+                                   guint nc_size, guint size,
+                                   clock_t *next_update_tick)
+{
+    clock_t now;
+    if (!label)
+        return;
+    now = clock ();
+    if (G_UNLIKELY (now > *next_update_tick))
+    {
+        gchar *text = g_strdup_printf ("%u/%u splits processed, %u combos",
+                                       nc_progress, nc_size, size);
+        status_update (label, text);
+        g_free (text);
+        *next_update_tick = now + MAX_UPDATE_TICKS;
+    }
+}
+
+class Hasher
+{
+public:
+    size_t operator() (gint64 const& key) const
+    {
+        return key;
+    }
+};
+class EqualFn
+{
+public:
+    bool operator() (gint64 const& t1, gint64 const& t2) const
+    {
+        return (t1 == t2);
+    }
+};
+
+gboolean
+gnc_autoclear_get_splits (Account *account, gnc_numeric toclear_value,
+                          time64 end_date,
+                          GList **splits, GError **error, GtkLabel *label)
+{
+    std::vector<Split*> nc_vector, empty_vec;
```

Comment:
Since you use `std::vector<Split*>` a lot you might want to make an alias for it:
```
using SplitVec = std::vector<Split*>;
```
then you can say
```
SplitVec nc_vector, empty_vec{}; // Always declare and initialize together.
```